File: video.c
/* --------------------- video.c -------------------- */

#include "dflat.h"
#include "termbox2.h"

#include "unikey.h"
#include <stdint.h>

int SCREENWIDTH = 80;
int SCREENHEIGHT = 24;
BOOL ClipString;

static const uintattr_t tb_base_colors[8] = {
    TB_BLACK, TB_BLUE, TB_GREEN, TB_CYAN,
    TB_RED, TB_MAGENTA, TB_YELLOW, TB_WHITE
};

static int in_screen(int x, int y)
{
    return x >= 0 && y >= 0 && x < SCREENWIDTH && y < SCREENHEIGHT;
}

uintattr_t tb_fg_from_attr(int attr)
{
    int fg = attr & 0x0f;
    uintattr_t color = tb_base_colors[fg & 7];

    if (fg & 8)
        color |= TB_BRIGHT;
    return color;
}

uintattr_t tb_bg_from_attr(int attr)
{
    return tb_base_colors[(attr >> 4) & 7];
}

static int tb_color_to_dflat(uintattr_t attr)
{
    switch (attr & 0xff) {
    case TB_BLACK:   return BLACK;
    case TB_BLUE:    return BLUE;
    case TB_GREEN:   return GREEN;
    case TB_CYAN:    return CYAN;
    case TB_RED:     return RED;
    case TB_MAGENTA: return MAGENTA;
    case TB_YELLOW:  return BROWN;
    case TB_WHITE:   return LIGHTGRAY;
    default:         return BLACK;
    }
}

static int dflat_attr_from_tb(uintattr_t fg, uintattr_t bg)
{
    int f = tb_color_to_dflat(fg);
    int b = tb_color_to_dflat(bg);

    if (fg & TB_BRIGHT)
        f += 8;
    return (f & 0x0f) | ((b & 7) << 4);
}

/* -- read a rectangle of video memory into a save buffer -- */
void getvideo(RECT rc, void far *bf)
{
    int ht = RectBottom(rc)-RectTop(rc)+1;
    int y;
    videocell_t far *p = bf;

    hide_mousecursor();
    for (y = RectTop(rc); y <= RectBottom(rc); y++) {
        int x;
        for (x = RectLeft(rc); x <= RectRight(rc); x++) {
            struct tb_cell *src = NULL;

            if (tb_get_cell(x, y, 1, &src) == TB_OK && src != NULL) {
                p->ch = src->ch;
                p->fg = src->fg;
                p->bg = src->bg;
            } else {
                p->ch = ' ';
                p->fg = TB_DEFAULT;
                p->bg = TB_DEFAULT;
            }
            p++;
        }
    }
    show_mousecursor();
}

/* -- write a rectangle of video memory from a save buffer -- */
void storevideo(RECT rc, void far *bf)
{
    int y;
    videocell_t far *p = bf;

    hide_mousecursor();
    for (y = RectTop(rc); y <= RectBottom(rc); y++) {
        int x;
        for (x = RectLeft(rc); x <= RectRight(rc); x++) {
            if (in_screen(x, y))
                tb_set_cell(x, y, p->ch, p->fg, p->bg);
            p++;
        }
    }
    show_mousecursor();
}

/* -------- read a character of video memory ------- */
videocell_t GetVideoChar(int x, int y)
{
    videocell_t c;
    struct tb_cell *src = NULL;

    hide_mousecursor();
    if (tb_get_cell(x, y, 1, &src) == TB_OK && src != NULL) {
        c.ch = src->ch;
        c.fg = src->fg;
        c.bg = src->bg;
    } else {
        c.ch = ' ';
        c.fg = TB_DEFAULT;
        c.bg = TB_DEFAULT;
    }
    show_mousecursor();
    return c;
}

/* -------- write a character of video memory ------- */
void PutVideoChar(int x, int y, videocell_t c)
{
    if (in_screen(x, y)) {
        hide_mousecursor();
        tb_set_cell(x, y, c.ch, c.fg, c.bg);
        show_mousecursor();
    }
}

BOOL CharInView(WINDOW wnd, int x, int y)
{
	WINDOW nwnd = NextWindow(wnd);
	WINDOW pwnd;
	RECT rc;
    int x1 = GetLeft(wnd)+x;
    int y1 = GetTop(wnd)+y;

	if (!TestAttribute(wnd, VISIBLE))
		return FALSE;
    if (!TestAttribute(wnd, NOCLIP))    {
        WINDOW wnd1 = GetParent(wnd);
        while (wnd1 != NULL)    {
            /* --- clip character to parent's borders -- */
			if (!TestAttribute(wnd1, VISIBLE))
				return FALSE;
			if (!InsideRect(x1, y1, ClientRect(wnd1)))
                return FALSE;
            wnd1 = GetParent(wnd1);
        }
    }
	while (nwnd != NULL)	{
		if (!isHidden(nwnd) /* && !isAncestor(wnd, nwnd) */ )	{
			rc = WindowRect(nwnd);
    		if (TestAttribute(nwnd, SHADOW))    {
        		RectBottom(rc)++;
        		RectRight(rc)++;
    		}
			if (!TestAttribute(nwnd, NOCLIP))	{
				pwnd = nwnd;
				while (GetParent(pwnd))	{
					pwnd = GetParent(pwnd);
					rc = subRectangle(rc, ClientRect(pwnd));
				}
			}
			if (InsideRect(x1,y1,rc))
				return FALSE;
		}
		nwnd = NextWindow(nwnd);
	}
    return (x1 < SCREENWIDTH && y1 < SCREENHEIGHT);
}

/* -------- write a character to a window ------- */
void wputch(WINDOW wnd, int c, int x, int y)
{
    if (CharInView(wnd, x, y))    {
        uint32_t ch = c & 0xffff;
        if (ch < 256)
            ch = kCp437[ch];
        int attr = clr(foreground, background);
        int xc = GetLeft(wnd)+x;
        int yc = GetTop(wnd)+y;

        hide_mousecursor();
        tb_set_cell(xc, yc, ch, tb_fg_from_attr(attr), tb_bg_from_attr(attr));
        show_mousecursor();
    }
}

/* ------- write a string to a window ---------- */
void wputs(WINDOW wnd, void *s, int x, int y)
{
    int x1 = GetLeft(wnd)+x;
    int x2 = x1;
    int y1 = GetTop(wnd)+y;
    if (x1 < SCREENWIDTH && y1 < SCREENHEIGHT && isVisible(wnd))	{
        videocell_t ln[MAXCOLS];
        videocell_t *cp1 = ln;
        unsigned char *str = s;
        int fg = foreground;
        int bg = background;
        int len;
        int off = 0;
        while (*str)    {
            if (*str == CHANGECOLOR)    {
                str++;
                foreground = (*str++) & 0x7f;
                background = (*str++) & 0x7f;
                continue;
            }
            if (*str == RESETCOLOR)    {
                foreground = fg & 0x7f;
                background = bg & 0x7f;
                str++;
                continue;
            }
            {
                uint32_t ch = (unsigned char)*str;
                if (*str == ('\t' | 0x80) || *str == ('\f' | 0x80))
                    ch = ' ';
                if (ch < 256 && (unsigned char)*str < 256)
                    ch = kCp437[(unsigned char)*str];
                cp1->ch = ch;
                cp1->fg = tb_fg_from_attr(clr(foreground, background));
                cp1->bg = tb_bg_from_attr(clr(foreground, background));
            }
            if (ClipString)
                if (!CharInView(wnd, x, y))
                    *cp1 = GetVideoChar(x2,y1);
            cp1++;
            str++;
            x++;
            x2++;
        }
        foreground = fg;
        background = bg;
        len = (int)(cp1-ln);
        if (x1+len > SCREENWIDTH)
            len = SCREENWIDTH-x1;

        if (!ClipString && !TestAttribute(wnd, NOCLIP))	{
            /* -- clip the line to within ancestor windows -- */
            RECT rc = WindowRect(wnd);
            WINDOW nwnd = GetParent(wnd);
            while (len > 0 && nwnd != NULL)	{
                if (!isVisible(nwnd))	{
                    len = 0;
                    break;
                }
                rc = subRectangle(rc, ClientRect(nwnd));
                nwnd = GetParent(nwnd);
            }
            while (len > 0 && !InsideRect(x1+off,y1,rc))	{
                off++;
                --len;
            }
            if (len > 0)	{
                x2 = x1+len-1;
                while (len && !InsideRect(x2,y1,rc))	{
                    --x2;
                    --len;
                }
            }
        }
        if (len > 0) {
            int i;
            hide_mousecursor();
            for (i = 0; i < len; i++)
                tb_set_cell(x1+off+i, y1, ln[off+i].ch, ln[off+i].fg, ln[off+i].bg);
            show_mousecursor();
        }
    }
}

/* --------- get the current video mode -------- */
void get_videomode(void)
{
    int w = tb_width();
    int h = tb_height();

    if (w > 0 && h > 0) {
        SCREENWIDTH = min(w, MAXCOLS - 1);
        SCREENHEIGHT = h - 1;
    }
}

void convert_screen_to_ansi()
{
    tb_present();
}

void scroll_window(WINDOW wnd, RECT rc, int d)
{
    int x1 = RectLeft(rc);
    int x2 = RectRight(rc);
    int y1 = RectTop(rc);
    int y2 = RectBottom(rc);
    int attr;
    int x, y;

    if (x1 < 0)
        x1 = 0;
    if (y1 < 0)
        y1 = 0;
    if (x2 >= SCREENWIDTH)
        x2 = SCREENWIDTH - 1;
    if (y2 >= SCREENHEIGHT)
        y2 = SCREENHEIGHT - 1;
    if (x1 > x2 || y1 >= y2)
        return;

    attr = clr(WndForeground(wnd), WndBackground(wnd));
    hide_mousecursor();
    if (d) {
        for (y = y1; y < y2; y++)
            for (x = x1; x <= x2; x++) {
                struct tb_cell *src = NULL;
                struct tb_cell cell;

                if (tb_get_cell(x, y + 1, 1, &src) == TB_OK && src != NULL) {
                    cell.ch = src->ch;
                    cell.fg = src->fg;
                    cell.bg = src->bg;
                } else {
                    cell.ch = ' ';
                    cell.fg = TB_DEFAULT;
                    cell.bg = TB_DEFAULT;
                }
                tb_set_cell(x, y, cell.ch, cell.fg, cell.bg);
            }
        for (x = x1; x <= x2; x++) {
            struct tb_cell cell;
            cell.ch = ' ';
            cell.fg = tb_fg_from_attr(attr);
            cell.bg = tb_bg_from_attr(attr);
            tb_set_cell(x, y2, cell.ch, cell.fg, cell.bg);
        }
    } else {
        for (y = y2; y > y1; y--)
            for (x = x1; x <= x2; x++) {
                struct tb_cell *src = NULL;
                struct tb_cell cell;

                if (tb_get_cell(x, y - 1, 1, &src) == TB_OK && src != NULL) {
                    cell.ch = src->ch;
                    cell.fg = src->fg;
                    cell.bg = src->bg;
                } else {
                    cell.ch = ' ';
                    cell.fg = TB_DEFAULT;
                    cell.bg = TB_DEFAULT;
                }
                tb_set_cell(x, y, cell.ch, cell.fg, cell.bg);
            }
        for (x = x1; x <= x2; x++) {
            struct tb_cell cell;
            cell.ch = ' ';
            cell.fg = tb_fg_from_attr(attr);
            cell.bg = tb_bg_from_attr(attr);
            tb_set_cell(x, y1, cell.ch, cell.fg, cell.bg);
        }
    }
    show_mousecursor();
}

File: src/Video.zig
const std = @import("std");
const df = @import("ImportC.zig").df;
const root = @import("root.zig");
const Window = @import("Window.zig");
const Rect = @import("Rect.zig");

const sTab:u16 = 0x0C + 0x80;

// assume video_address is 16 bites (2 bytes)
fn vad(x:usize, y:usize) usize {
    return y * @as(usize, @intCast(df.SCREENWIDTH)) + x;
}

fn movetoscreen(bf:[]u16, offset:usize, len:usize) void {
    // video_address is u8
    const bf8:[]u8 = @ptrCast(bf);
    @memcpy(df.video_address[offset*2..(offset+len)*2], bf8[0..len*2]);
}

fn movefromscreen(bf:[]u16, offset:usize, len:usize) void {
    // video_address is u8
    const bf8:[]u8 = @ptrCast(bf);
    @memcpy(bf8[0..len*2], df.video_address[offset*2..(offset+len)*2]);
}

// -- read a rectangle of video memory into a save buffer --
pub fn getvideo(rc:df.RECT, bf:[]u16) void {
    const ht:usize = @intCast(rc.bt-rc.tp+1);
    const bytes_row:usize = @intCast(rc.rt-rc.lf+1);
    var vadr = vad(@intCast(rc.lf), @intCast(rc.tp));
    df.hide_mousecursor();
    for (0..ht) |idx| {
        movefromscreen(bf[bytes_row*idx..], vadr, bytes_row);
        vadr += @as(usize, @intCast(df.SCREENWIDTH));
    }
    df.show_mousecursor();
}

// -- write a rectangle of video memory from a save buffer --
pub fn storevideo(rc:df.RECT, bf:[]u16) void {
    const ht:usize = @intCast(rc.bt-rc.tp+1);
    const bytes_row:usize = @intCast(rc.rt-rc.lf+1);
    var vadr = vad(@intCast(rc.lf), @intCast(rc.tp));
    df.hide_mousecursor();
    for (0..ht) |idx| {
        movetoscreen(bf[bytes_row*idx..], vadr, bytes_row);
        vadr += @intCast(df.SCREENWIDTH);
    }
    df.show_mousecursor();
}

// -------- read a character of video memory -------
pub fn GetVideoChar(x:usize, y:usize) u16 {
    df.hide_mousecursor();
    // #define peek(a,o)       (*((unsigned short *)((char *)(a)+(o))))
    // const c = peek(video_address, vad(x,y));
    const va16_ptr:[*]u16 = @ptrCast(@alignCast(df.video_address));
    const c:[*]u16 = va16_ptr+vad(x,y);
    df.show_mousecursor();
    return c[0];
}

// -------- write a character of video memory -------
pub fn PutVideoChar(x:usize, y:usize, chr:u16) void {
    if (x < df.SCREENWIDTH and y < df.SCREENHEIGHT) {
        df.hide_mousecursor();
        // #define poke(a,o,w)     (*((unsigned short *)((char *)(a)+(o))) = (w))
        // poke(video_address, vad(x,y), chr);
        const va16_ptr:[*]u16 = @ptrCast(@alignCast(df.video_address));
        const c:[*]u16 = va16_ptr+vad(x,y);
        c[0] = chr;
        df.show_mousecursor();
    }
}

pub fn CharInView(win:*Window, x:usize, y:usize) bool {
    const left:usize = win.GetLeft();
    const top:usize = win.GetTop();
    const x1:usize = left+x;
    const y1:usize = top+y;

    if (win.TestAttribute(df.VISIBLE) == false)
        return false;

    if (win.TestAttribute(df.NOCLIP) == false) {
        var ww = win.parent;
        while (ww) |win1| {
            // --- clip character to parent's borders --
            if (win1.TestAttribute(df.VISIBLE) == false)
                return false;
            if (Rect.InsideRect(x1, y1, win1.ClientRect()) == false)
                return false;
            ww = win1.parent;
        }
    }

    var nwin = win.nextWindow();
    while (nwin) |nw| {
        if (nw.isHidden() == false) { //  && !isAncestor(wnd, nwnd)
            var rc = nw.WindowRect();
            if (nw.TestAttribute(df.SHADOW)) {
                rc.bottom += 1;
                rc.right += 1;
            }
            if (nw.TestAttribute(df.NOCLIP) == false) {
                var pp = nw;
                while (pp.parent) |pwin| {
                    pp = pwin;
                    rc = Rect.subRectangle(rc, pwin.ClientRect());
                }
            }
            if (Rect.InsideRect(x1,y1,rc))
                return false;
        }
         nwin = nw.nextWindow();
    }
    return (x1 < df.SCREENWIDTH and y1 < df.SCREENHEIGHT);
}

// -------- write a character to a window -------
pub fn wputch(win:*Window, chr:u16, x:usize, y:usize) void {
    if (CharInView(win, x, y)) {
        // #define clr(fg,bg) ((fg)|((bg)<<4))
        const ch:u16 = @intCast((chr & 255) | ((df.foreground | (df.background << 4)) << 8));
        const xc:usize = win.GetLeft()+x;
        const yc:usize = win.GetTop()+y;
        df.hide_mousecursor();
        // #define poke(a,o,w)     (*((unsigned short *)((char *)(a)+(o))) = (w))
        // poke(video_address, vad(xc, yc), ch);
        const va16_ptr:[*]u16 = @ptrCast(@alignCast(df.video_address));
        const c:[*]u16 = va16_ptr+vad(xc,yc);
        c[0] = ch;
        df.show_mousecursor();
    }
}

// ------- write a string to a window ----------
pub fn wputs(win:*Window, s:[:0]const u8, x:usize, y:usize) void {
    const x1 = win.GetLeft()+x;
    const y1 = win.GetTop()+y;
    var xx = x;
    var x2 = x1;
    var idx:usize = 0;
    var ln = [_]u16{0}**df.MAXCOLS;
    var ldx:usize = 0;
    const fg = df.foreground;
    const bg = df.background;
    if (x1 < df.SCREENWIDTH and y1 < df.SCREENHEIGHT and win.isVisible()) {
        while ((idx < s.len) and (s[idx] != 0)){
            if (s[idx] == df.CHANGECOLOR) {
                df.foreground = s[idx+1] & 0x7f;
                df.background = s[idx+2] & 0x7f;
                idx += 3;
                continue;
            }
            if (s[idx] == df.RESETCOLOR) {
                df.foreground = fg & 0x7f;
                df.background = bg & 0x7f;
                idx += 1;
                continue;
            }
            if (s[idx] == ('\t' | 0x80) or s[idx] == (sTab | 0x80)) {
                ln[ldx] =  @intCast(' ' | ((df.foreground | (df.background << 4)) << 8));
            } else {
                ln[ldx] = @intCast((s[idx] & 255) | ((df.foreground | (df.background << 4)) << 8));
                if (df.ClipString>0) {
                    if (CharInView(win, xx, y)  == false) {
                        const va16_ptr:[*]u16 = @ptrCast(@alignCast(df.video_address));
                        const c:[*]u16 = va16_ptr+vad(x2,y1);
                        ln[ldx] = c[0];
                    }
                }
            }
            idx += 1;
            ldx += 1;
            xx += 1;
            x2 += 1;
        }
        df.foreground = fg;
        df.background = bg;
        var len = ldx;
        if (x1+len > df.SCREENWIDTH)
            len = @as(usize, @intCast(df.SCREENWIDTH))-x1;

        var off:usize = 0;
        if (df.ClipString == 0 and win.TestAttribute(df.NOCLIP) == false) {
            // -- clip the line to within ancestor windows --
            var rc = win.WindowRect();
            var nwnd = win.parent;
            while (len > 0 and nwnd != null) {
                const nwin = nwnd.?;
                if (nwin.isVisible() == false) {
                    len = 0;
                    break;
                }
                rc = Rect.subRectangle(rc, nwin.ClientRect());
                nwnd = nwin.parent;
            }
            while (len > 0 and Rect.InsideRect(x1+off,y1,rc) == false) {
                off += 1;
                len -|= 1;
            }
            if (len > 0) {
                x2 = x1+len-1;
                while (len>0 and Rect.InsideRect(x2,y1,rc) == false) {
                    x2 -|= 1;
                    len -|= 1;
                }
            }
        }
        if (len > 0) {
            df.hide_mousecursor();
            movetoscreen(ln[off..], vad(x1+off,y1), len);
            df.show_mousecursor();
        }

    }
}

// --------- get the current video mode --------
pub fn get_videomode() void {
    if (df.video_address == null)
        df.video_address = df.tty_allocate_screen(df.SCREENWIDTH, df.SCREENHEIGHT);
}
